<a href="https://colab.research.google.com/github/lauquintero28/Business-Analitics/blob/main/Reto2_Modelo%C3%81rbol.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**0. Cargar las librerías.**

In [ ]:
import numpy as np #Funciones matemáticas
import pandas as pd #Conexión con Archivos de Excel
import matplotlib.pyplot as plt #Gráficas

#Se trae el modelo
from sklearn.tree import DecisionTreeClassifier #Arbol de Decisión - Prestamo Interbibliotecario

#Para llamar los archivos del Drive - Sin tener que subirlos
from google.colab import drive #Archivos
drive.mount('/content/drive')

Mounted at /content/drive


##**1. Cargar la base de datos del drive.**

In [ ]:
nxl='/content/drive/MyDrive/EAFIT/TERCER SEMESTRE/Analítica de Negocios/2. BD2_In_App_Pur_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.dropna() #Evitar celdas vacias Not at Number
XDB.head()

#Creamos el DataMart
XD=XDB[['Age','Income','Years','Hours']] #Variables Entrada
XD.head()
XD=np.array(XD) #Convierto la tabla en números - Variables de Entrada
yd=XDB[['Buy']] #Valor Deseado - Aprenda el Modelo
yd=np.array(yd) #Variable de Salida - Convierto en Tabla

#**2. Se procede con la construcción del Modelo de Árbol.**

In [ ]:
mar=DecisionTreeClassifier(criterion="gini",max_depth=4)
mar.fit(XD,yd)
ydp=mar.predict(XD) # Estos valores de pronóstico sirven para evaluar modelo

#Evaluamos el modelo
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

EX=(VP+VN)/(VP+VN+FP+FN)
Se=VP/(VP+FN); Sp=VN/(VN+FP)

print("Exactitud: ",EX)
print("Sensibilidad: ",Se)
print("Especificidad: ",Sp)

Exactitud:  0.8475
Sensibilidad:  0.9478672985781991
Especificidad:  0.7354497354497355


#**3. Se procede con la visualización del Árbol.**


In [ ]:
from sklearn.tree import export_graphviz   # Estructura del Árbol
from pydotplus import graph_from_dot_data  # Que datos le voy a colocar al árbol

vs=["Age","Income","Years","Hours"] # Estos nombres aparecen en el árbol

dot_data=export_graphviz(mar,feature_names=vs) # Asegurar nombres del gráfico
graph=graph_from_dot_data(dot_data) # Se hace el gráfico del árbol
graph.write_png("arbol.png") # Se crea la imagen

True

#**4. Calculamos las probabilidades de los compradores nuevos.**

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2=XDB2[["Age","Income","Years","Hours"]]
XDB2=XDB2.dropna()

XD2=np.array(XDB2)
ydp2=mar.predict(XD2)
XDB2["Buy"]=ydp2
XDB2.head()

XDB2.to_excel("Compradores.xlsx")

# Para contar los datos
np.sum(ydp2)
len(ydp2)-np.sum(ydp2)

1

#**Análisis de resultados.**

En la base de datos se registran un total de **400** jugadores, de los cuales **211** **(53%)** son compradores potenciales y **189** **(47%)** no realizan compras dentro de la aplicación. El modelo de árbol de decisión mostró un índice de **exactitud** de aproximadamente **85%**, indicando un buen rendimiento en la clasificación general. Además, el modelo exhibe una **sensibilidad** cercana al **95%**, lo que refleja su alta capacidad para identificar correctamente a los compradores potenciales. Por otro lado, el índice de **especificidad** es aproximadamente **74%**, midiendo la efectividad del modelo para clasificar correctamente a los jugadores que no efectuarán compras. Estos índices nos demuestran la capacidad del modelo para distinguir entre compradores y no compradores dentro de la aplicación.

De acuerdo con los nuevos casos que agrupa la base de datos In_App_Pur_Socre, la probabilidad de que el primer jugador haga compras dentro de la aplicación es del **67%** y la probabilidad para el segundo jugador es del **0%.**


Las reglas para los nodos puros son:

- La regla para el nodo puro [5,0] es:
SI Years<0.5 AND Age>23 AND Years <=1.5 AND Income>76.5

- La regla para el nodo puro [28,0] es:
SI Income>63.5 AND Years<=1.5 AND Years>0.5

- La regla para el nodo puro [90,0] es:
SI Income<=63.5


Las reglas para los nuevos casos son:

- Income>63.5 - Falso - 128
- Years>1.5 - Falso - 3
- Years<=0.5 - Verdadero
- Income>76.5 - Falso
